### PR TITLE
DSG-667: Add additional user fields and simplify create endpoint.

### PR DESCRIPTION
**Pull request by JayStGelais**
This change removes the "inital roles" collection from the create user request. It is expected that automated systems that will handle user sync from external identitiy providers will not know what roles exist in a system. Instead we will later implement a concpet to configure default role sets given to users based on `identityProvider` and/or `userType`. 

For now, roles will need to be assigned manually after first login.

Additionally, this change adds `identityProvider` and`userType` to the user record to support future features that will depend on these. The uniqueness constraint for `externalId` has been updated to take `identityProvider` into consideration to better support deployments where multiple identity providers are in use.

## How Has This Been Tested?
 - Unit tests updated to reflect new intended behaviors.
 - Manual testing in Postman to exercise unique external user checking as this requires postgres SQL schema.

## Types of changes

Patch
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation change (documentation only, will not require redeployment)
- [ ] Refactor or cleanup (functionality unchanged)
Minor
- [ ] New feature (non-breaking change which adds functionality)
Major
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have performed a self-review of my own code
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been merged and published in downstream modules

- [x] I have confirmed all acceptance criteria defined by the story have been met

### CODE REVIEW

File: gradle/code-quality.gradle [hunk 1]
```diff
@@ -70,7 +70,7 @@ jacocoTestCoverageVerification {
     violationRules {
         rule {
             limit {
-                minimum = 0.81
+                minimum = 0.8
```
Comment (JayStGelais) [inline gradle/code-quality.gradle:73]: Note: I delete a lot of previous covered code as part of this change. I did add a new, previously missing test in the area impacted, but still failed to hit the old threshold that we were barely at.

We have upcoming planned work to address existing unit test gaps.

File: service/build.gradle [hunk 2]
```diff
@@ -46,7 +46,7 @@ dependencies {
 
 	// Database management
 	implementation 'org.liquibase:liquibase-core'
-	runtimeOnly 'org.postgresql:postgresql:42.5.1'
+	implementation 'org.postgresql:postgresql:42.5.1'
```
Comment (deibys-nuvalence) [inline service/build.gradle:49]: why this change is required ?

File: service/src/main/java/io/nuvalence/user/management/api/service/entity/UserEntityEventListener.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+ * Automatically sets created at timestamp on users.
+ */
+public class UserEntityEventListener {
+    /**
+     * Sets created at timestamp.
+     *
+     * @param user userbeing persisted
+     */
+    @PrePersist
+    public void preUpdateTrackedEntityPersist(final UserEntity user) {
```
Comment (deibys-nuvalence) [inline service/src/main/java/io/nuvalence/user/management/api/service/entity/UserEntityEventListener.java:17]: is user parameter used ?

File: service/src/main/java/io/nuvalence/user/management/api/service/controller/UsersApiDelegateImpl.java [hunk 4]
```diff
@@ -188,4 +209,15 @@ private Page<UserEntity> handleRequestsWithRolesNames(
...
         }
     }
+
+    private boolean isDuplicateExternalUser(final DataIntegrityViolationException e) {
+        return ExceptionInspectionUtility.findCauseOfType(
+                        e, org.hibernate.exception.ConstraintViolationException.class)
+                .map(org.hibernate.exception.ConstraintViolationException.class::cast)
+                .map(
+                        cve ->
+                                "user_table_identity_provider_external_id_key"
```
Comment (deibys-nuvalence) [inline service/src/main/java/io/nuvalence/user/management/api/service/controller/UsersApiDelegateImpl.java:219]: This looks very coupled with the name of an object in db .....before  creating user cannot we verify if user already exist ? may be an additional query but may make things cleaner

File: service/src/main/java/io/nuvalence/user/management/api/service/entity/UserEntity.java [hunk 5]
```diff
@@ -42,24 +44,32 @@
...
+@SuppressWarnings("ClassFanOutComplexity")
 public class UserEntity {
     @Id
     @GeneratedValue(generator = "uuid2")
     @Type(type = "org.hibernate.type.PostgresUUIDType")
     @Column(name = "id", updatable = false, nullable = false)
     private UUID id;
 
-    @Column(name = "external_id")
+    @Column(name = "external_id", nullable = false)
```
Comment (deibys-nuvalence) [inline service/src/main/java/io/nuvalence/user/management/api/service/entity/UserEntity.java:56]: Should this change be reflected also in db ?
I see this in liquibase file
`<column name="external_id" type="varchar(255)">
                <constraints nullable="true" unique="true" />
            </column>`

File: service/src/main/java/io/nuvalence/user/management/api/service/enums/UserType.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+import com.fasterxml.jackson.annotation.JsonValue;
+
+/**
+ * Enumeration of user types.
+ */
+public enum UserType {
+    Agency("agency"),
+    Public("public"),
+    Unknown("unknown");
```
Comment (deibys-nuvalence) [inline service/src/main/java/io/nuvalence/user/management/api/service/enums/UserType.java:11]: JUst for my knowledge , what is the use case of "unknown" ?

File: service/src/main/resources/db/liquibase-changelog.xml [hunk 7]
```diff
@@ -78,4 +78,20 @@
...
+        <!-- These constraints don't get created in H2 databases used for unit testing -->
+        <sql dbms="postgresql">
+            ALTER TABLE user_table DROP CONSTRAINT IF EXISTS user_table_email_key;
+            ALTER TABLE user_table DROP CONSTRAINT IF EXISTS user_table_external_id_key;
+        </sql>
+        <addColumn tableName="user_table">
+            <column name="identity_provider" type="varchar(1024)">
+                <constraints nullable="false" />
+            </column>
+            <column name="user_type" type="varchar(255)">
```
Comment (deibys-nuvalence) [inline service/src/main/resources/db/liquibase-changelog.xml:91]: minor , can be ignored . Isnt 255 length too high for values like "agency" , "unknown", "public" ?

File: service/src/main/resources/db/liquibase-changelog.xml [hunk 8]
```diff
@@ -78,4 +78,20 @@
         <addPrimaryKey tableName="user_role" columnNames="user_id, role_id"/>
         <addPrimaryKey tableName="application_permission" columnNames="application_id, permission_id"/>
     </changeSet>
+    <changeSet id="add-usertype-and-identityproider-to-user_table" author="jstgelais">
+        <!-- These constraints don't get created in H2 databases used for unit testing -->
+        <sql dbms="postgresql">
+            ALTER TABLE user_table DROP CONSTRAINT IF EXISTS user_table_email_key;
```
Comment (deibys-nuvalence) [inline service/src/main/resources/db/liquibase-changelog.xml:84]: Cannot these sql statements be declared using liquibase syntax ? like in https://docs.liquibase.com/change-types/drop-unique-constraint.html

File: service/build.gradle (hunk 2 above)
Comment (JayStGelais) [reply service/build.gradle:49]: It's not. Good catch. When I was trying to detect the unique constraint violation I was going down the route of thinking I needed the Postgres Exception, which was the most root cause of the failure.

File: service/src/main/resources/db/liquibase-changelog.xml (hunk 8 above)
Comment (JayStGelais) [reply service/src/main/resources/db/liquibase-changelog.xml:84]: They could, but because H2 doesn't build unique constraints, they don't exist when initializing unit tests using H2. So this is a worek around to get this to only apply to postgres.

File: service/src/main/resources/db/liquibase-changelog.xml (hunk 7 above)
Comment (JayStGelais) [reply service/src/main/resources/db/liquibase-changelog.xml:91]: I went with a value that is relatively small in storage terms but big enough to accomodate future use cases without having to resize it later.

File: service/src/main/java/io/nuvalence/user/management/api/service/enums/UserType.java (hunk 6 above)
Comment (JayStGelais) [reply service/src/main/java/io/nuvalence/user/management/api/service/enums/UserType.java:11]: This was to gracefully handle cases where new values end up in the API without the enum being update. It's used as the default case in mapping strings to this enum.

File: service/src/main/java/io/nuvalence/user/management/api/service/entity/UserEntity.java (hunk 5 above)
Comment (JayStGelais) [reply service/src/main/java/io/nuvalence/user/management/api/service/entity/UserEntity.java:56]: Good catch. For some reason I thought this field was already non-nullable in the database and I thought I was brining the JPA configuraion in line with it. I'll add this to liquibase.

File: service/src/main/resources/db/liquibase-changelog.xml (hunk 8 above)
Comment (deibys-nuvalence) [reply service/src/main/resources/db/liquibase-changelog.xml:84]: ok

File: service/src/main/java/io/nuvalence/user/management/api/service/controller/UsersApiDelegateImpl.java (hunk 4 above)
Comment (deibys-nuvalence) [reply service/src/main/java/io/nuvalence/user/management/api/service/controller/UsersApiDelegateImpl.java:219]: My comment is more targeted at how we check for user uniqueness here . 
As I understand we are saving the User object through using JPA which hides all implementation to connect and issue queries to the database , but in case the constraint  `user_table_identity_provider_external_id_key` is not satisfied we check in service code if in the exception comes how the constraint is named and that is what I consider odd to check for constraint database names in our code.

I may be missing something but there is something else I don't understand . In this same PR you are dropping the constrain ` ALTER TABLE user_table DROP CONSTRAINT IF EXISTS user_table_external_id_key;` , so if constraint does not exist anymore potentially when would this exception be raised ?

**Review by deibys-nuvalence — APPROVED** [review]
Approved with agreed changes in the comments

File: gradle/code-quality.gradle (hunk 1 above)
Comment (JayStGelais) [reply gradle/code-quality.gradle:73]: The most recent commit raised coverage back to 81 percent, but I would prefer we leave this on a round number. This way if the next code refactor that comes across removes unnecessary code, we aren't stuck against this tight limit.

**Review by deibys-nuvalence — APPROVED** [review]
LGTM
